### DISCUSSION TITLE
[DropdownMenu] Mobile/responsive friendly version

### DISCUSSION THREAD

**Discussion by rafgraph** [radix-ui/primitives#796] [Help]
Hi, I really like the potential of the DropdownMenu component, but it easily runs out of screen real estate on small screens. This happens with sub-menus, long menus, and when using a mobile device in landscape mode, etc (also adding `overflow-y: scroll` to `DropdownMenu.Content` doesn't make make the menu content scrollable).

In general I don't think the overlay pattern is well suited for mobile devices because of these issues (unless the content is only a few items).

On small screens is it possible to show the menu not as an overlay? That is, either expand the page down to show the menu items, or have it take over the whole screen when it's opened? Also, the sub menu would need to expand down, instead of to the side.

If it's not currently possible, it'd be nice if you'd create a responsive version of the DropdownMenu. To take this a step further, it'd be awesome if you separated the menu content from the menu trigger type. With the trigger type being dropdown, context, expand page, whole screen, something custom, etc, so that the same menu content can be shown in different ways depending on the screen size and positioning on the page (i.e. responsive).

![IMG_1598](https://user-images.githubusercontent.com/11911299/126809147-091d8e85-171b-4ba5-a5ec-be8655d02e49.jpeg)

**Comment by benoitgrelard** [comment]
Hi @rafgraph, thanks for bringing this up.

Yeah, this is a tricky one. With our primitives, we try to stay as style-agnostic as possible. We don't really want to have opinions on that front, as much as possible. Most of the things mentioned here "mostly" about styling. I say "mostly" because the line between styling and intrinsic functionality can sometimes be blurry.

This is definitely something we need to explore more though. We've discussed this a little with the team, and we generally don't think "responsive" primitives is the way to go. We think we'd rather expose some more specific "mobile" parts which people can swap at will.

So for example, we may decide to expose a `DropdownMenu.ContentTray` or something of sorts, and submenus would stack on top of each other, and you'd get a `DropdownMenu.BackButton` part to navigate back to a parent menu.

Another temporary solution might be that we unlock our primitives even more in some areas. We try to keep them as open as possible by exposing all the parts but there are still some bits that aren't exposed. For examples positioning wrappers/portals, etc. If we were to expose those, you would be able to override them to suit your needs better.

This is all just ideas in our heads at the moment, as we have had other priorities to pursue for now but this is most definitely something we would like to solve at some point.

**Reply by rafgraph** [reply]
**Is it possible to "style" the dropdown menu to expand the page instead of render as an overlay/popover? How do you recommend implementing/styling long dropdown menus with sub menus on mobile?** I was unable to figure out a workable solution.

I think we're kind of saying the same thing with regards to responsive. I like the mobile parts idea that the engineer can swap at will. The engineer can use the different parts (and trigger types?) to create the desired responsive experience, but not that primitives themselves would have the responsiveness pre-defined (I agree that doesn't make sense).

**Reply by benoitgrelard** [reply]
> Is it possible to "style" the dropdown menu to expand the page instead of render as an overlay/popover? How do you recommend implementing/styling long dropdown menus with sub menus on mobile? I was unable to figure out a workable solution.

Yeah this is what I meant when I said:

> Another temporary solution might be that we unlock our primitives even more in some areas. We try to keep them as open as possible by exposing all the parts but there are still some bits that aren't exposed. For examples positioning wrappers/portals, etc. If we were to expose those, you would be able to override them to suit your needs better.

Meaning that right now, there is no easy way to get out of the layering + positioning logic.

**Reply by rafgraph** [reply]
That sounds great. The more declarative (and less imperative) it can be the better! All of my projects require feature parity across desktop and mobile so this is much needed.

**Comment by g1na1011** [comment]
Hi there! I was wondering if there's been any progress on the proposed ideas that @benoitgrelard had mentioned previously? Or if not, @rafgraph - were you able to turn your dropdown menus to be mobile-friendly? I am struggling to figure out how to get the `DropdownMenu.Content` to take over the entire screen when we hit smaller breakpoints since I don't think we can currently override the portal's styling. Does it make sense to use a combination of the `DropdownMenu` and the `Dialog` (assuming we can have the dialog take up the whole screen) to handle mobile situations?

**Comment by thewebartisan7** [comment]
Hi there,

I stumbled upon the same problem, so I would like to share my solution, and ask if there are any progress about.

Regarding the main Content height, I used the ScrollArea component to add and set a max height (thanks to `--radix-popper-available-height`) on the main content:

```js
    <DropdownMenu.Content>
      <ScrollArea.Root>
        {children}
      </ScrollArea.Root>
    </DropdownMenu.Content>
```

Then for sub content, I just hide it when in mobile and then show another Dropdown.Group which then is hidden on desktop.

So I made some new prop to show/hide menu items and groups of items depending on breakpoints.

But I like the idea of `DropdownMenu.ContentTray` and `DropdownMenu.BackButton` mentioned by @benoitgrelard 

I have some issue with sub content, when it's being open and I scroll up/down, then it doesn't close sub content, but I didn't yet check into how to close programmatically maybe when scroll happen (like for example stop scroll area when sub menu is open by setting overflow hidden?), see:

![6ac8164e520d3cd77718150c725f4cca](https://user-images.githubusercontent.com/57477934/235733467-4bc967ba-4078-452e-b71b-0ce1bd091a08.gif)

**Reply by skarensmoll** [reply]
HI @thewebartisan7 were you able to fix the latest problem? "when it's being open and I scroll up/down, then it doesn't close sub content" would be nice if you could share your solution :)
